docs/_posts/2024-07-18-Run-your-pandas-workloads-on-cluster.md

```diff
@@ -0,0 +1,202 @@
+
+### Workload
+
+For the experiments, we will be using a script that involves three operations: `read_csv`, `apply`, and `map`.
+These operations have been selected to minimize data transfer between nodes, ensuring that the primary
+focus is on executing the workload efficiently within a cluster. The aim is to optimize performance by
+leveraging cluster capabilities without the added complexity of data movement between nodes.
+
+<details>
+  <summary>pandas_workload.py</summary>
+
+  <pre><code class="language-python">
+    import pandas as pd
+    file_path = "/home/ray/data/big_yellow.csv"
+    df = pd.read_csv(file_path)
+    df = df.map(str)
+    result = pd.DataFrame()
+
+    payment_type_map = {
+        1: 'Credit Card',
+        2: 'Cash',
+        3: 'No Charge',
+        4: 'Dispute',
+        5: 'Unknown',
+        6: 'Voided Trip'
+    }
+
+    result["df_payment_type"] = df['payment_type'].map(payment_type_map)
+
+    def custom_function(row):
+        return str(row["passenger_count"]) + " passengers were picked up at " + str(row["tpep_pickup_datetime"])
+
+    result["description"] = df.apply(custom_function, axis=1)
+
+  </code></pre>
+</details>
+
+### Running script with pandas 
+
+To run this workload, we chose a t3.2xlarge instance with 32GB of RAM, enough to hold all the data in memory.
+However, when processing the script, pandas encountered an Out of Memory (OOM) error and crashed
+because the processing dataframe required more memory than a single node could provide.
+The original data fits into memory, but the additional memory needed during processing caused the issue.
+Since a single node was insufficient for the workload, we had two options: scale horizontally
+(using multiple nodes) or vertically (using a more powerful machine). pandas, limited to a single node,
+can't scale horizontally. Fortunately, Modin on Ray can distribute the workload across multiple nodes,
+making horizontal scaling possible.
+
+### Prerequisites
+
+Modin makes it a breeze to run your existing pandas code in a cluster with minimal changes.
+You can simply submit your Python script as a Ray job to a Ray cluster. Just keep in mind a couple of prerequisites:
+
+1.  You need to have a Ray cluster up and running.
+2.  The dataset must be available (not just accessible) on all nodes in the cluster.
+
+#### Setting Up Ray Cluster and Distributing the Dataset
+
+There are several ways to set up a Ray cluster. You can opt for an on-premise setup or
+take the easier route of configuring a cluster in AWS using tools like
+[KubeRay](https://github.com/ray-project/kuberay/tree/master/helm-chart/ray-cluster).
+In our configuration, we used [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
+to download and prepare data on all nodes of the Ray cluster, ensuring that both prerequisites are effectively met.
+
+Once the Ray cluster is up and running, set the  `RAY_ADDRESS` environment variable
```

Comment:
Do we really need this?

Reply:
I thought this made sense, as it tells the user how to make use of the remote ray cluster.

Reply:
This is related to a Ray instance to connect to but you mention the Ray dashboard URL. Should we put away words about the dashboard?

Reply:
To submit a job to a ray cluster the ray dashboard url has to be set for RAY_ADDRESS as mentioned here https://docs.ray.io/en/latest/cluster/running-applications/job-submission/quickstart.html#submitting-a-job